File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function(root, subRoot) {
    if(!subRoot){
        return true
    }
    if(!root){
        return false
    }
    if(isSameTree(root, subRoot)){
        return true
    }

    return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot)
};

function isSameTree(r, sub){
    if(!r && !sub){
        return true
    }
    if(r && sub && r.val == sub.val){
        return isSameTree(r.left, sub.left) && isSameTree(r.right, sub.right)
    }
    return false
}

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        root_path = self.preorder(root)
        subRoot_path = self.preorder(subRoot)

        return subRoot_path in root_path

    def preorder(self, node):
        if not node:
            return "N"
        path = ""
        path += "," + str(node.val) + ","
        path += self.preorder(node.left)
        path += self.preorder(node.right)
        return path
        